packages/slice-machine/src/ErrorBoundary.tsx

```diff
@@ -0,0 +1,43 @@
+import { ErrorBoundary as EditorUiErrorBoundary } from "@prismicio/editor-ui";
+import {
+  isUnauthenticatedError,
+  isUnauthorizedError,
+} from "@slicemachine/manager/client";
+import { type ComponentProps, type FC, useCallback, useRef } from "react";
+
+import { useAuthStatus } from "@src/hooks/useAuthStatus";
+
+type ErrorBoundaryProps = Pick<
+  EditorUiErrorBoundaryProps,
+  "children" | "renderError"
+>;
+// TODO(DT-1979): Export the `ErrorBoundaryProps` type from `@prismicio/editor-ui`.
+type EditorUiErrorBoundaryProps = ComponentProps<typeof EditorUiErrorBoundary>;
+
+export const ErrorBoundary: FC<ErrorBoundaryProps> = (props) => {
+  const errorRef = useRef<unknown>();
+  const authStatus = useAuthStatus();
+  return (
+    <EditorUiErrorBoundary
+      {...props}
+      onError={(error) => {
+        errorRef.current =
+          isUnauthenticatedError(error) || isUnauthorizedError(error)
+            ? error
+            : undefined;
+      }}
+      ref={useCallback(
+        (errorBoundary: EditorUiErrorBoundary | null) => {
+          if (errorBoundary === null || errorRef.current === undefined) return;
+          errorRef.current = undefined;
+          errorBoundary.reset();
+        },
+        /* eslint-disable-next-line react-hooks/exhaustive-deps --
+         * Whenever `authStatus` changes, we want to `reset` the `errorBoundary`
+         * if an authentication or authorization error has been caught.
+         **/
+        [authStatus],
+      )}
```

Comment:
Hm, it still feels like a clever use of `ref`, in the non-obvious way. It requires detailed knowledge of React.

We are trying to react upon a change in `authState`, not the re-render of ErrorBoundary, which is why I feel attaching the reset logic to `authState` makes more sense. The linked example where a DOM element's height is saved as a side-effect seems more fitting since it is reacting to the re-render of that `ref`ed element.

You are right that `onChange` + `useEffect` feels a bit inelegant. Event handlers aren't always a good option.

Here's a modified version that contains all logic within ErrorBoundary. The previous proposed changes to `useAuthStatus` can be ignored.

```tsx
import { ErrorBoundary as EditorUiErrorBoundary } from "@prismicio/editor-ui";
import { type ComponentProps, type FC, useRef } from "react";

import { useAuthStatus } from "@src/hooks/useAuthStatus";

// TODO(DT-1979): Export the `ErrorBoundaryProps` type from `@prismicio/editor-ui`.
type ErrorBoundaryProps = ComponentProps<typeof EditorUiErrorBoundary>;

export const ErrorBoundary: FC<ErrorBoundaryProps> = (props) => {
  const errorBoundary = useRef<EditorUiErrorBoundary>(null);
  const authStatus = useAuthStatus();

  const previousAuthStatus = useRef(authStatus);
  if (previousAuthStatus.current !== authStatus) {
    errorBoundary.current?.reset();
  }
  previousAuthStatus.current = authStatus;

  return <EditorUiErrorBoundary {...props} ref={errorBoundary} />;
};
```

This implementation ensures the Suspense cache is cleared on any auth status change, not just ones picked up by an ErrorBoundary, and more specifically, ones that correctly throw `UnauthorizedError` or `UnauthenticatedError`. It changes original intent of this PR slightly, but it might be more correct. Please feel free to challenge that assumption if you think it's incorrect.

I'll leave it at that and let you decide. Both implementations should get the job done. 🙂

Should we stick with the current SM-specific ErrorBoundary implementation, is it possible to retain the ability to use a custom `onError` and `ref`? It could be as simple as calling `props.onError` and `props.ref` within the already present callbacks.